src/net_processing.cpp

```diff
@@ -3563,32 +3563,44 @@ void ProcessMessage(
     return;
 }
 
+/** Maybe disconnect a peer and discourage future connections from its address.
+ *
+ * @param[in]   pnode           The node to check.
+ * @return                      True if the peer will be disconnected.
+ */
 bool PeerLogicValidation::MaybeDiscourageAndDisconnect(CNode& pnode)
 {
-    LOCK(cs_main);
-    CNodeState &state = *State(pnode.GetId());
+    {
+        LOCK(cs_main);
+        CNodeState &state = *State(pnode.GetId());
+
+        // There's nothing to do if the m_should_discourage flag isn't set
+        if (!state.m_should_discourage) return false;
 
-    if (state.m_should_discourage) {
+        // Reset m_should_discourage
         state.m_should_discourage = false;
-        if (pnode.HasPermission(PF_NOBAN)) {
-            LogPrintf("Warning: not punishing whitelisted peer %s!\n", pnode.addr.ToString());
-        } else if (pnode.m_manual_connection) {
-            LogPrintf("Warning: not punishing manually-connected peer %s!\n", pnode.addr.ToString());
-        } else if (pnode.addr.IsLocal()) {
-            // Disconnect but don't discourage this local node
-            LogPrintf("Warning: disconnecting but not discouraging local peer %s!\n", pnode.addr.ToString());
-            pnode.fDisconnect = true;
-        } else {
-            // Disconnect and discourage all nodes sharing the address
-            LogPrintf("Disconnecting and discouraging peer %s!\n", pnode.addr.ToString());
-            if (m_banman) {
-                m_banman->Discourage(pnode.addr);
-            }
-            connman->DisconnectNode(pnode.addr);
-        }
+    } // cs_main
+
+    if (pnode.HasPermission(PF_NOBAN)) {
+        // Log but don't disconnect
+        LogPrintf("Warning: not punishing whitelisted peer %s!\n", pnode.addr.ToString());
+        return false;
+    } else if (pnode.m_manual_connection) {
```

Comment:
37e7265796102df35ec83b53fe9114e07d214910

Maybe drop these `else` since you early return on each block.

Reply:
Agree. That's much clearer.